File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays.DotNet/director/Scores/RaysScoreReader.cs
using Microsoft.Extensions.Logging;
using ProjectorRays.Common;
using ProjectorRays.director.Scores.Data;
using static ProjectorRays.director.Scores.RayScoreTagsV2;
using static ProjectorRays.director.Scores.RaysScoreChunk;

namespace ProjectorRays.director.Scores;
internal class RaysScoreReader
{
    private readonly ILogger _logger;

   

    public RaysScoreReader(ILogger logger)
    {
        _logger = logger;
    }


    #region Header
    internal void ReadAllIntervals(int entryCount, ReadStream stream, RayScoreParseContext ctx)
    {
        var s = new ReadStream(new BufferView(stream.Data, stream.Offset, stream.Size),
            stream.Endianness, stream.Pos, Annotator);

        int[] offsets = new int[entryCount + 1];
        for (int i = 0; i < offsets.Length; i++)
            offsets[i] = s.ReadInt32($"offset[{i}]");

        int entriesStart = s.Pos;

        if (entryCount < 1)
            return;

        var size = offsets[1] - offsets[0];
        int absoluteStart = stream.Offset + entriesStart + offsets[0];

        ctx.SetFrameDataBufferView(stream.Data, absoluteStart, size);
        List<int> intervalOrder = new();
        if (entryCount >= 2)
        {
            size = offsets[2] - offsets[1];
            int absoluteStart2 = stream.Offset + entriesStart + offsets[1];
            var orderView = new BufferView(stream.Data, absoluteStart2, offsets[2] - offsets[1]);
            var os = new ReadStream(orderView, Endianness.BigEndian, annotator: Annotator);
            if (os.Size >= 4)
            {
                int count = os.ReadInt32("orderCount");
                for (int i = 0; i < count && os.Pos + 4 <= os.Size; i++)
                    intervalOrder.Add(os.ReadInt32("order", new() { ["index"] = i }));
            }
        }
        
        var entryIndices = intervalOrder.Count > 0 ? intervalOrder : null;
        if (entryIndices == null)
        {
            entryIndices = new List<int>();
            for (int i = 3; i + 2 < offsets.Length; i += 3)
                entryIndices.Add(i);
        }
        ctx.ResetFrameDescriptorBuffers();

        foreach (int primaryIdx in entryIndices)
        {
            if (primaryIdx + 2 >= offsets.Length)
                continue;

            size = offsets[primaryIdx + 1] - offsets[primaryIdx];
            int absoluteStart2 = stream.Offset + entriesStart + offsets[primaryIdx];
            ctx.AddFrameDescriptorBuffer(new BufferView(stream.Data, absoluteStart2, size));


            var secSize = offsets[primaryIdx + 2] - offsets[primaryIdx + 1];
            if (secSize > 0)
            {
                int absoluteStart3 = stream.Offset + entriesStart + offsets[primaryIdx + 1];
                var secView = new BufferView(stream.Data, absoluteStart3, secSize);
                ctx.AddBehaviorScriptBuffer(secView);
            }
        }
        
    }
    internal RayScoreHeader ReadMainHeader(ReadStream stream)
    {
        return new RayScoreHeader
        {
            TotalLength = stream.ReadInt32(),
            HeaderType = stream.ReadInt32(), // constantMinus3
            OffsetsOffset = stream.ReadInt32(), // constant12
            EntryCount = stream.ReadInt32(),
            NotationBase = stream.ReadInt32(), // entryCountPlus1
            EntrySizeSum = stream.ReadInt32(),
        };
    }

    internal void ReadHeader(ReadStream stream, RayScoreHeader header)
    {
        header.ActualSize = stream.ReadInt32("actualSize");
        header.UnkA1 = stream.ReadUint8("unkA1");
        header.UnkA2 = stream.ReadUint8("unkA2");
        header.UnkA3 = stream.ReadUint8("unkA3");
        header.UnkA4 = stream.ReadUint8("unkA4");
        header.HighestFrame = stream.ReadInt32("highestFrame");
        header.UnkB1 = stream.ReadUint8("unkB1");
        header.UnkB2 = stream.ReadUint8("unkB2");
        header.SpriteSize = stream.ReadInt16("spriteSize");
        header.UnkC1 = stream.ReadUint8("unkC1");
        header.UnkC2 = stream.ReadUint8("unkC2");
        header.ChannelCount = stream.ReadInt16("channelCount");
        //header.FirstBlockSize = stream.ReadInt16("firstBlockSize");
    }


    #endregion



    public ScoreTagMain? ReadMainTag(ushort data, ReadStream stream,RayScoreParseContext ctx)
    {
        var tag = RayScoreTagsV2.TryParseTagMain(data);
        return tag;
    }


    public void ApplyKnownTag(RaySprite sprite, ScoreTagV2 tag, byte[] data, RayScoreParseContext ctx)
    {
        var rs = new ReadStream(data, data.Length, Endianness.BigEndian, annotator: ctx.Annotator);
        int expectedSize = GetDataLength(tag);
        if (expectedSize > 0 && data.Length != expectedSize)
            ctx.Logger.LogWarning($"Tag {tag:X4} expected {expectedSize} bytes, got {data.Length}");
        _logger.LogInformation($"Sprite Tag parse:{tag}:{string.Join(',', data.Select(x => $"{x.ToString("X2")}({x})"))}");
        switch (tag)
        {
            case ScoreTagV2.Ease:
                sprite.EaseIn = rs.ReadUint8("easeIn", ctx.GetAnnotationKeys());
                sprite.EaseOut = rs.ReadUint8("easeOut", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.Curvature:
                sprite.Curvature = rs.ReadUint16("curv", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.AdvanceFrame:
                rs.ReadUint16("adv", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.Flags:
                rs.ReadUint16("flags", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.FlagsControl:
                rs.ReadUint16("flagsCtrl", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.TweenFlags:
                sprite.TweenFlags = ParseTweenFlags(rs.ReadUint8("tweenFlags", ctx.GetAnnotationKeys()));
                break;
            default:
                break;
        }
    }
    
    
    public void ApplyKnownTag(RayScoreKeyFrame kf, ScoreTagV2 tag, byte[] data, RayScoreParseContext ctx)
    {
        var rs = new ReadStream(data, data.Length, Endianness.BigEndian, annotator: ctx.Annotator);
        int expectedSize = GetDataLength(tag);
        if (expectedSize > 0 && data.Length != expectedSize)
            ctx.Logger.LogWarning($"Tag {tag:X4} expected {expectedSize} bytes, got {data.Length}");
        _logger.LogInformation($"KeyFrame Tag parse:{tag}:{string.Join(',', data.Select(x => $"{x.ToString("X2")}({x})"))}");
        switch (tag)
        {
            case ScoreTagV2.Size:
                kf.Width = rs.ReadInt16("width", ctx.GetAnnotationKeys());
                kf.Height = rs.ReadInt16("height", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.Position:
                kf.LocH = rs.ReadInt16("locH", ctx.GetAnnotationKeys());
                kf.LocV = rs.ReadInt16("locV", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.Colors:
                kf.ForeColor = rs.ReadUint8("foreColor", ctx.GetAnnotationKeys());
                kf.BackColor = rs.ReadUint8("backColor", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.Ink:
                kf.Ink = rs.ReadInt16("ink", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.Rotation:
                kf.Rotation = rs.ReadInt16("rot", ctx.GetAnnotationKeys()) / 100f;
                break;
            case ScoreTagV2.Skew:
                kf.Skew = rs.ReadInt16("skew", ctx.GetAnnotationKeys()) / 100f;
                break;
            case ScoreTagV2.AdvanceFrame:
                rs.ReadUint16("adv", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.Composite:
                kf.Width = rs.ReadInt16("width", ctx.GetAnnotationKeys());
                kf.Height = rs.ReadInt16("height", ctx.GetAnnotationKeys());
                byte blendRaw = rs.ReadUint8("blendRaw", ctx.GetAnnotationKeys());
                kf.Blend = 100 - blendRaw * 100 / 255;
                kf.Ink = rs.ReadUint8("ink", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.FrameRect:
                kf.LocH = rs.ReadInt16("locH", ctx.GetAnnotationKeys());
                kf.LocV = rs.ReadInt16("locV", ctx.GetAnnotationKeys());
                kf.Width = rs.ReadInt16("width", ctx.GetAnnotationKeys());
                kf.Height = rs.ReadInt16("height", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.Flags:
                rs.ReadUint16("flags", ctx.GetAnnotationKeys());
                break;
            case ScoreTagV2.FlagsControl:
                rs.ReadUint16("flagsCtrl", ctx.GetAnnotationKeys());
                break;
            default:
                break;
        }
    }
   
    public RaySprite CreateChannelSprite(ReadStream stream, RayScoreParseContext ctx)
    {
        var startPos = stream.Pos;
        var sprite = new RaySprite();
        var keys = ctx.GetAnnotationKeys();
        //var flags1 = stream.ReadUint8();

        byte flagByte = stream.ReadUint8("tweenFlags", keys);
        sprite.TweenFlags = ParseTweenFlags(flagByte);
        
        var something2 = stream.ReadInt16();

        //_logger.LogInformation("Sprite flags=" + flags1);

        //if ((flags1 & 0xFF) != 0 && (flags1 & 0x10) == 0)
        //{
        //    var keyframeType = stream.ReadUint8();
        //    var keyframeSize = stream.ReadUint8();
        //    _logger.LogInformation($"something1={keyframeType}, something2={keyframeSize}");
        //}

        byte inkByte = stream.ReadUint8("ink", keys);
        sprite.Ink = inkByte & 0x7F;
        sprite.ForeColor = stream.ReadUint8("foreColor", keys);
        sprite.BackColor = stream.ReadUint8("backColor", keys);
        sprite.MemberCastLib = stream.ReadUint16("memberCastLib", keys);
        sprite.MemberNum = stream.ReadUint16("memberNum", keys);
        stream.Skip(2);
        sprite.SpritePropertiesOffset = stream.ReadUint16("propOffset", keys);
        sprite.LocV = stream.ReadInt16("locV", keys);
        sprite.LocH = stream.ReadInt16("locH", keys);
        sprite.Height = stream.ReadInt16("height", keys);
        sprite.Width = stream.ReadInt16("width", keys);
        byte colorcode = stream.ReadUint8("color", keys);
        sprite.Editable = (colorcode & 0x40) != 0;
        sprite.ScoreColor = colorcode & 0x0F;
        var blend = stream.ReadUint8("blendRaw", keys);
        sprite.Blend = (int)Math.Round(100f - blend / 255f * 100f);
        byte flag2 = stream.ReadUint8("flipFlags", keys);
        sprite.FlipV = (flag2 & 0x04) != 0;
        sprite.FlipH = (flag2 & 0x02) != 0;
        stream.Skip(5);
        sprite.Rotation = stream.ReadInt32("rotation", keys) / 100f;
        sprite.Skew = stream.ReadInt32("skew", keys) / 100f;
        var skip = 48 - (stream.Pos - startPos);
        stream.Skip(skip); // We need to read exactly 48 bytes

        _logger.LogInformation($"{sprite.LocH}x{sprite.LocV}:Ink={sprite.Ink}:Blend={sprite.Blend}:Skew={sprite.Skew}:Rot={sprite.Rotation}:PropOffset={sprite.SpritePropertiesOffset}:Member={sprite.MemberCastLib},{sprite.MemberNum}:Tween={ sprite.TweenFlags}");
        return sprite;
    }


    public RayTweenFlags ParseTweenFlags(byte value)
    {
        return new RayTweenFlags
        {
            TweeningEnabled = (value & 0x01) != 0,
            Path = (value & 0x02) != 0,
            Size = (value & 0x04) != 0,
            Rotation = (value & 0x08) != 0,
            Skew = (value & 0x10) != 0,
            Blend = (value & 0x20) != 0,
            ForeColor = (value & 0x40) != 0,
            BackColor = (value & 0x80) != 0
        };
    }


    #region Frame descriptors

    internal RayScoreIntervalDescriptor? ReadFrameIntervalDescriptor(int index, ReadStream stream)
    {
        if (stream.Size < 44)
            return null;

        var desc = new RayScoreIntervalDescriptor();
        var k = new Dictionary<string, int> { ["entry"] = index };
        desc.StartFrame = stream.ReadInt32("startFrame", k);
        desc.EndFrame = stream.ReadInt32("endFrame", k);
        desc.Unknown1 = stream.ReadInt32("unk1", k);
        desc.Unknown2 = stream.ReadInt16("unk2", k);

        // Correctly cast to flag enum and extract bitfield
        var flags = (RaySpriteFlags)stream.ReadInt16("spriteFlagsBitfield", k);

        desc.FlipH = flags.HasFlag(RaySpriteFlags.FlipH);
        desc.FlipV = flags.HasFlag(RaySpriteFlags.FlipV);
        desc.Editable = flags.HasFlag(RaySpriteFlags.Editable);
        desc.Moveable = flags.HasFlag(RaySpriteFlags.Moveable);
        desc.Trails = flags.HasFlag(RaySpriteFlags.Trails);
        desc.IsLocked = flags.HasFlag(RaySpriteFlags.Locked);


        desc.Channel = stream.ReadInt32("channel", k);
        desc.UnknownAlwaysOne = stream.ReadInt16("const1", k);
        desc.UnkownA = stream.ReadInt16("unkownA", k);
        desc.UnkownB = stream.ReadInt16("unkownB", k);
        desc.UnknownE1 = stream.ReadUint8("e1", k);
        desc.UnknownFD = stream.ReadUint8("fd", k);
        desc.Unknown7 = stream.ReadInt16("unk7", k);
        desc.Unknown8 = stream.ReadInt32("unk8", k);
        while (stream.Pos + 4 <= stream.Size)
            desc.ExtraValues.Add(stream.ReadInt32("extra", k));
        _logger.LogInformation($"Item Desc. {index}: Start={desc.StartFrame}, End={desc.EndFrame}, Channel={desc.Channel}, U1={desc.Unknown1}, Flip={desc.FlipH},{desc.FlipV},🔒={desc.IsLocked},Trails={desc.Trails},Editable={desc.Editable},Moveable={desc.Moveable} , U3={desc.UnknownAlwaysOne}, U4A={desc.UnkownA}, U4B={desc.UnkownB}, U5={desc.UnknownE1}, U6={desc.UnknownFD}");
        return desc;
    }
    internal void ReadFrameDescriptors(RayScoreParseContext ctx)
    {
        ctx.ResetFrameDescriptors();

        int ind = 0;
        foreach (var frameIntervalDescriptor in ctx.FrameIntervalDescriptorBuffers)
        {
            var ps = new ReadStream(frameIntervalDescriptor, Endianness.BigEndian,
                annotator: Annotator);
            var descriptor = ReadFrameIntervalDescriptor(ind, ps);
            if (descriptor != null)
                ctx.AddFrameDescriptor(descriptor);


            ind++;
        }
    }
    #endregion



    #region Behaviors
    internal void ReadBehaviors(RayScoreParseContext ctx)
    {
        int ind = 0;
        foreach (var frameIntervalDescriptor in ctx.BehaviorScriptBuffers)
        {
            var ps = new ReadStream(frameIntervalDescriptor, Endianness.BigEndian,
                annotator: Annotator);
            var behaviourRefs = ReadBehaviors(ind, ps);
            if (ind < ctx.Descriptors.Count)
                ctx.Descriptors[ind].Behaviors.AddRange(behaviourRefs);
            ctx.AddFrameScript(behaviourRefs);

            ind++;
        }
    }
    internal static List<RaysBehaviourRef> ReadBehaviors(int ind, ReadStream ss)
    {
        var behaviours = new List<RaysBehaviourRef>();
        while (ss.Pos + 8 <= ss.Size)
        {
            var k = new Dictionary<string, int> { ["entry"] = ind };
            short cl = ss.ReadInt16("castLib", k);
            short cm = ss.ReadInt16("castMmb", k);
            ss.ReadInt32("zero", k);
            behaviours.Add(new RaysBehaviourRef { CastLib = cl, CastMmb = cm });
        }
        return behaviours;
    }

    internal void ApplyKnownTag(object currentSprite, ScoreTagV2 value, byte[] data, RayScoreParseContext ctx)
    {
        throw new NotImplementedException();
    }

    #endregion



}

File: WillMoveToOwnRepo/ProjectorRays/src/ProjectorRays.Python/director/scores/rays_score_reader.py
from ..chunks.rays_chunk import RaysChunk, ChunkType
from ..common.stream import ReadStream, Endianness
from .data import RaySprite, RayScoreHeader, RayScoreIntervalDescriptor, RayScoreKeyFrame, RayTweenFlags
from . import ray_score_tags_v2 as tags
from .ray_sprite_factory import RaySpriteFactory

class RaysScoreReader:
    """Very small subset reader for score chunks."""
    def __init__(self, logger=None):
        self.logger = logger

    def read_main_header(self, stream: ReadStream) -> RayScoreHeader:
        header = RayScoreHeader()
        header.total_length = stream.read_uint32()
        header.header_type = stream.read_uint32()
        header.offsets_offset = stream.read_uint32()
        header.entry_count = stream.read_uint32()
        header.notation_base = stream.read_uint32()
        header.entry_size_sum = stream.read_uint32()
        return header

    def parse_tween_flags(self, value: int) -> RayTweenFlags:
        return RayTweenFlags(
            tweening_enabled=bool(value & 0x01),
            path=bool(value & 0x02),
            size=bool(value & 0x04),
            rotation=bool(value & 0x08),
            skew=bool(value & 0x10),
            blend=bool(value & 0x20),
            fore_color=bool(value & 0x40),
            back_color=bool(value & 0x80),
        )

    def create_channel_sprite(self, stream: ReadStream) -> RaySprite:
        sp = RaySprite()
        sp.tween_flags = self.parse_tween_flags(stream.read_uint8())
        stream.read_uint16()  # unknown
        sp.ink = stream.read_uint8()
        sp.fore_color = stream.read_uint8()
        sp.back_color = stream.read_uint8()
        sp.member_cast_lib = stream.read_uint16()
        sp.member_num = stream.read_uint16()
        stream.read_uint16()
        sp.sprite_properties_offset = stream.read_uint16()
        sp.loc_v = stream.read_int16()
        sp.loc_h = stream.read_int16()
        sp.height = stream.read_int16()
        sp.width = stream.read_int16()
        sp.editable = bool(stream.read_uint8() & 0x40)
        sp.blend = 100 - int(stream.read_uint8() * 100 / 255)
        flip = stream.read_uint8()
        sp.flip_v = bool(flip & 0x04)
        sp.flip_h = bool(flip & 0x02)
        stream.read_bytes(5)
        sp.rotation = stream.read_int32() / 100.0
        sp.skew = stream.read_int32() / 100.0
        return sp

    def read_main_tag(self, data: int, stream: ReadStream, ctx):
        return tags.try_parse_tag_main(data)

    # ------------------------------------------------------------------
    # Additional methods ported from the C# implementation
    # ------------------------------------------------------------------

    def read_all_intervals(self, entry_count: int, stream: ReadStream, ctx):
        from ..common.stream import BufferView
        s = ReadStream(BufferView(stream.data, 0, len(stream.data)), stream.endianness, stream.pos)
        offsets = [s.read_int32() for _ in range(entry_count + 1)]
        entries_start = s.pos
        if entry_count < 1:
            return
        size = offsets[1] - offsets[0]
        absolute_start = entries_start + offsets[0]
        ctx.set_frame_data_buffer_view(stream.data, absolute_start, size)
        interval_order = []
        if entry_count >= 2:
            size = offsets[2] - offsets[1]
            absolute_start2 = entries_start + offsets[1]
            order_view = BufferView(stream.data, absolute_start2, offsets[2] - offsets[1])
            os = ReadStream(order_view, Endianness.BIG)
            if len(os.data) >= 4:
                count = os.read_int32()
                for i in range(count):
                    if os.pos + 4 <= len(os.data):
                        interval_order.append(os.read_int32())

        entry_indices = interval_order if interval_order else None
        if entry_indices is None:
            entry_indices = []
            i = 3
            while i + 2 < len(offsets):
                entry_indices.append(i)
                i += 3
        ctx.reset_frame_descriptor_buffers()
        for primary_idx in entry_indices:
            if primary_idx + 2 >= len(offsets):
                continue
            size = offsets[primary_idx + 1] - offsets[primary_idx]
            absolute_start2 = entries_start + offsets[primary_idx]
            ctx.add_frame_descriptor_buffer(BufferView(stream.data, absolute_start2, size))
            sec_size = offsets[primary_idx + 2] - offsets[primary_idx + 1]
            if sec_size > 0:
                absolute_start3 = entries_start + offsets[primary_idx + 1]
                ctx.add_behavior_script_buffer(BufferView(stream.data, absolute_start3, sec_size))

    def read_header(self, stream: ReadStream, header: RayScoreHeader):
        header.actual_size = stream.read_int32()
        header.unk_a1 = stream.read_uint8()
        header.unk_a2 = stream.read_uint8()
        header.unk_a3 = stream.read_uint8()
        header.unk_a4 = stream.read_uint8()
        header.highest_frame = stream.read_int32()
        header.unk_b1 = stream.read_uint8()
        header.unk_b2 = stream.read_uint8()
        header.sprite_size = stream.read_int16()
        header.unk_c1 = stream.read_uint8()
        header.unk_c2 = stream.read_uint8()
        header.channel_count = stream.read_int16()

    def read_frame_interval_descriptor(self, index: int, stream: ReadStream):
        if len(stream.data) < 44:
            return None
        desc = RayScoreIntervalDescriptor()
        desc.start_frame = stream.read_int32()
        desc.end_frame = stream.read_int32()
        desc.unknown1 = stream.read_int32()
        desc.unknown2 = stream.read_int16()
        flags = stream.read_int16()
        desc.flip_h = bool(flags & 0x01)
        desc.flip_v = bool(flags & 0x40)
        desc.editable = bool(flags & 0x20)
        desc.moveable = bool(flags & 0x10)
        desc.trails = bool(flags & 0x08)
        desc.is_locked = bool(flags & 0x04)

        desc.channel = stream.read_int32()
        desc.unknown_always_one = stream.read_int16()
        desc.unkown_a = stream.read_int16()
        desc.unkown_b = stream.read_int16()
        desc.unknown_e1 = stream.read_uint8()
        desc.unknown_fd = stream.read_uint8()
        desc.unknown7 = stream.read_int16()
        desc.unknown8 = stream.read_int32()
        while stream.pos + 4 <= len(stream.data):
            desc.extra_values.append(stream.read_int32())
        return desc

    def read_frame_descriptors(self, ctx):
        ctx.reset_frame_descriptors()
        for ind, buf in enumerate(ctx.frame_interval_descriptor_buffers):
            ps = ReadStream(buf, Endianness.BIG)
            descriptor = self.read_frame_interval_descriptor(ind, ps)
            if descriptor is not None:
                ctx.add_frame_descriptor(descriptor)

    def read_behaviors(self, ctx):
        for ind, buf in enumerate(ctx.behavior_script_buffers):
            ps = ReadStream(buf, Endianness.BIG)
            behaviour_refs = self.read_behaviors_block(ind, ps)
            if ind < len(ctx.descriptors):
                ctx.descriptors[ind].behaviors.extend(behaviour_refs)
            ctx.add_frame_script(behaviour_refs)

    def read_behaviors_block(self, ind: int, stream: ReadStream):
        from .data import RaysBehaviourRef
        behaviours = []
        while stream.pos + 8 <= len(stream.data):
            cl = stream.read_int16()
            cm = stream.read_int16()
            stream.read_int32()
            behaviours.append(RaysBehaviourRef(cl, cm))
        return behaviours
    def apply_known_tag_sprite(self, sprite: RaySprite, tag: tags.ScoreTagV2, data: bytes):
        rs = ReadStream(data, Endianness.BIG)
        if tag == tags.ScoreTagV2.Ease:
            sprite.ease_in = rs.read_uint8()
            sprite.ease_out = rs.read_uint8()
        elif tag == tags.ScoreTagV2.Curvature:
            sprite.curvature = rs.read_uint16()
        elif tag == tags.ScoreTagV2.TweenFlags:
            sprite.tween_flags = self.parse_tween_flags(rs.read_uint8())

    def apply_known_tag_keyframe(self, kf: RayScoreKeyFrame, tag: tags.ScoreTagV2, data: bytes):
        rs = ReadStream(data, Endianness.BIG)
        if tag == tags.ScoreTagV2.Size:
            kf.width = rs.read_int16()
            kf.height = rs.read_int16()
        elif tag == tags.ScoreTagV2.Position:
            kf.loc_h = rs.read_int16()
            kf.loc_v = rs.read_int16()
        elif tag == tags.ScoreTagV2.Colors:
            kf.fore_color = rs.read_uint8()
            kf.back_color = rs.read_uint8()
        elif tag == tags.ScoreTagV2.Ink:
            kf.ink = rs.read_int16()
        elif tag == tags.ScoreTagV2.Rotation:
            kf.rotation = rs.read_int16() / 100.0
        elif tag == tags.ScoreTagV2.Skew:
            kf.skew = rs.read_int16() / 100.0
        elif tag == tags.ScoreTagV2.Composite:
            kf.width = rs.read_int16()
            kf.height = rs.read_int16()
            blend_raw = rs.read_uint8()
            kf.blend = 100 - int(blend_raw * 100 / 255)
            kf.ink = rs.read_uint8()
        elif tag == tags.ScoreTagV2.FrameRect:
            kf.loc_h = rs.read_int16()
            kf.loc_v = rs.read_int16()
            kf.width = rs.read_int16()
            kf.height = rs.read_int16()

